### ISSUE TITLE
#pragma support (OpenMP specifically)

### ISSUE THREAD

**Issue by bgoodri** [stan-dev/stan#439]
1. Amend the parser to not throw away anything to the right (inclusive) of `#pragma ...`
2. There is no step 2.

**Comment by syclik** [comment]
@bgoodri, is there a concrete suggestion you have? Is this something we need to code generate? Is this something that should go down in the math library? Maybe in the distributions?

**Comment by bgoodri** [comment]
The first was "don't drop comments that start with a #", in which case users could do their own #pragmas. But users would still have to add `-include omp.h -fopenmp` to get it to work, and could crash their program (which I don't care much about) if they try to do something in parallel that has to be serial. The second was to introduce a `pfor (j in 1:J)` block that would code generate to 
```
#pragma omp for
for (j in 1:J) {
  // do stuff that refers to j
}
```
in which case we would at least know to generate a line with `#include omp.h` and possibly we could do the for loop in a safer way. Sebastian was playing around with different `#pragmas` that would preclude some problems in Stan.

I doubt there is any distribution in Stan Math with accumulated analytical gradients where OpenMP would be worth the overhead. And Bob is suggesting that distributions without accumulated analytical gradients would be problematic for OpenMP. Eigen is doing some stuff like this for operations on giant matrices.

Also, none of the OpenMP stuff works on a Mac with Xcode. You have to install clang or g++ from upstream.

**Comment by syclik** [comment]
Let's start with the first thing you mentioned: "don't drop comments that start with a #". Can you tell me if I'm right in my interpretation?
- deprecate `#` as a comment character
- directly copy lines that start with `#` into the generated C++ code
If so, can you create a new issue that clearly states that? If I'm mistaken, can you correct me?

Regarding OpenMP, @rtrangucci and I actually got a version of clang++ installed (I forget how now, but it's not Xcode) and showed a little speed boost on a super-large matrix multiplication, but that's neither here nor there.

Should we close this issue and create two issues:
1. something to do with `#` as directives and not comments
2. creating a `pfor` function in the Stan language?

**Comment by bgoodri** [comment]
> If so, can you create a new issue that clearly states that?
That is literally the OP on this issue, but it is an open question as to whether that is a good idea. The `pfor` thing could be a separate issue, but we should probably do one or the other rather than both.

**Comment by syclik** [comment]
The OP wasn't clear enough for me.

If you think we should do one and not both, please help us make that decision. What are the pros and cons of doing one or the other? What's your preference? Who's actually going to implement it and what's that person's preference?

**Comment by bob-carpenter** [comment]
I suggest we also move this to longer-term issues
until there's more of a design as to what we should do.

We don't need to decide who's going to implement something
when we create an issue---it just has to be precise enough
to be implemented.

For the OpenMP pragma, I don't think there's any way we
could parallelize Stan code naively without synchronizing
on the expression graph for autodiff.  That would presumably
require the pieces that are being executed in parallel to
not touch the expression graph, which means analytical derivatives.

The case Sebastian was demo-ing ran multiple ODE solvers
in parallel.  And I think he had a way to synch the results
(maybe another comment-like thing?)
